### PR TITLE
Implements rint

### PR DESCRIPTION
**Pull request by khurd21**
Functions implemented:

- ccm::rint(float)
- ccm::rint(double)
- ccm::rint(long double)
- ccm::rintl(long double)
- ccm::rintf(float)

- ccm::lrint(float)
- ccm::lrint(double)
- ccm::lrint(long double)
- ccm::lrintl(long double)
- ccm::lrintf(float)

- ccm::llrint(float)
- ccm::llrint(double)
- ccm::llrint(long double)
- ccm::llrintl(long double)
- ccm::llrintf(long double)

Tests written for:

- Common and edge case conditions for `float`, `double`, and `long
  double` inputs.
- Assert that FE_INVALID is raised for +/- infinity values and NaN
  values.
- Asserts that functions can be evaluated at compile time.

### CODE REVIEW

File: include/ccmath/math/nearest/rint.hpp [hunk 1]
```diff
@@ -8,7 +8,117 @@
...
+	template <class T, std::enable_if_t<std::is_floating_point_v<T>, bool> = true>
+	constexpr T rint(T num) noexcept
+	{
+		if (ccm::isinf(num) || ccm::isnan(num)) { ccm::support::fenv::raise_except_if_required(FE_INVALID); }
+		constexpr auto rounding_mode{ccm::support::fenv::get_rounding_mode()};
+		return ccm::support::fp::directional_round(num, rounding_mode);
+	}
+
+	template <class Integer, std::enable_if_t<std::is_integral_v<Integer>, bool> = true>
+	constexpr double rint(Integer num)
```
Comment (khurd21) [inline include/ccmath/math/nearest/rint.hpp:28]: Add noexcept? Also need to document each function.

Comment (Rinzii) [reply include/ccmath/math/nearest/rint.hpp:28]: Following the defined API of the standard there should not be a `noexcept` for ccm::rint. This may change later in the future but for the time being aim to mimic the API pretty closely.

**Comment by Rinzii** [comment]
Glad to see a new PR! I will warn you though. I've recently pushed a major internal restructure of the project to the `dev` branch so make sure your fork is up to date!

File: include/ccmath/math/nearest/rint.hpp (hunk 1 above)
Comment (Rinzii) [reply include/ccmath/math/nearest/rint.hpp:28]: You can also use this as a reference:

https://en.cppreference.com/w/cpp/numeric/math/rint

Comment (khurd21) [reply include/ccmath/math/nearest/rint.hpp:44]: @Rinzii Sounds good. I removed the noexcept for that method!

I think it is ready for a review!

File: test/nearest/rint_test.cpp [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+
+	template <typename InputType, typename OutputType = InputType>
+	struct RIntTestParams
+	{
+		InputType input{};
+		OutputType output{};
+	};
+
+	template <typename T>
+	std::vector<T> makeTestParams()
```
Comment (khurd21) [inline test/nearest/rint_test.cpp:38]: After pushing this all up, I realized I had a question: Is there a specific casing I should be following? I am assuming snake case to match the standard library?

Comment (Rinzii) [reply test/nearest/rint_test.cpp:38]: There is no hard rule, but generally yes I follow pretty close to LLVMs style using snake case. You can also use clang-format to aid you in making sure your following the formatting style of the project!

File: include/ccmath/math/nearest/rint.hpp [hunk 3]
```diff
@@ -10,7 +10,239 @@
...
+	 * overloads of std::rint for all cv-unqualified floating-point types as the type of the parameter num.
+	 * @tparam T The type of the number.
+	 * @param num A floating-point value.
+	 * @return If no errors occur, the nearest integer value to num, according to the current rounding mode, is returned.
+	 */
+	template <class T, std::enable_if_t<std::is_floating_point_v<T>, bool> = true>
+	constexpr T rint(T num)
+	{
+		if (ccm::isinf(num) || ccm::isnan(num)) { ccm::support::fenv::raise_except_if_required(FE_INVALID); }
+		constexpr auto rounding_mode{ccm::support::fenv::get_rounding_mode()};
```
Comment (Rinzii) [inline include/ccmath/math/nearest/rint.hpp:33]: This function can't have `rounding_mode` as `constexpr` because if the compiler evaluates `get_rounding_mode()` at compile time, it will lock in a specific rounding mode (e.g., round-to-nearest) regardless of the actual rounding mode at runtime. This would lead to incorrect rounding behavior when the function is called at runtime since the rounding mode may differ from the one determined at compile time. So I'd remove the `constexpr` qualifier here.

Comment (Rinzii) [reply include/ccmath/math/nearest/rint.hpp:33]: Of course you can change this to `const` to take advantage of constant folding!

File: include/ccmath/math/nearest/rint.hpp [hunk 4]
```diff
@@ -10,7 +10,239 @@
...
+
+	/**
+	 * @brief Rounds the floating-point argument num to an integer value, using the current rounding mode. The library provides overloads of std::lrint and
+	 * std::llrint for all cv-unqualified floating-point types as the type of the parameter num.(
+	 * @param num A floating-point value.
+	 * @return If no errors occur, the nearest integer value to num, according to the current rounding mode, is returned.
+	 */
+	constexpr long lrint(float num)
+	{
+		return static_cast<long>(ccm::rint(num));
```
Comment (Rinzii) [inline include/ccmath/math/nearest/rint.hpp:136]: Ok, this function here is a bit trickier and performing a `static_cast` won't actually be enough to respect all rounding modes. `static_cast` does not perform conversion with floats but instead performs truncation and this will always follow the default rounding mode. As such if the rounding mode is different than the default this should not work as expected. The proper solution is to instead get the bits of the float itself and perform rounding towards a signed integer using the bits themselve. There is a few manners to do this, but it will be a tricky problem to fix. Let me know your thoughts!

File: include/ccmath/math/nearest/rint.hpp [hunk 5]
```diff
@@ -10,7 +10,239 @@
...
+
+	/**
+	 * @brief Rounds the floating-point argument num to an integer value, using the current rounding mode. The library provides overloads of std::lrint and
+	 * std::llrint for all cv-unqualified floating-point types as the type of the parameter num.(
+	 * @param num A floating-point value.
+	 * @return If no errors occur, the nearest integer value to num, according to the current rounding mode, is returned.
+	 */
+	constexpr long lrint(double num)
+	{
+		if (ccm::isnan(num))
```
Comment (Rinzii) [inline include/ccmath/math/nearest/rint.hpp:147]: This will also suffer from the issue I mentioned earlier with `static_cast` along with every other `lrint` function that converts a float to a integer.

**Review by Rinzii — CHANGES_REQUESTED** [review]
Thank you for taking the time to contribute to CCMath! The work that people like you do is invaluable and I greatly appreciate the time you have put in.

Overall, this is a good start! I have a few minor change requests, but also some major change requests that may be trickier to deal with. If you need help or advice on some of the requests I've made feel free to ping me.

I look forward to your changes!

File: test/nearest/rint_test.cpp (hunk 2 above)
Comment (khurd21) [reply test/nearest/rint_test.cpp:38]: Sounds good, thanks. I set up the clang format in VSCode to run on every save, however it didn't seem to adjust for the snake vs camel case. I will update the casing for this commit! :)

File: test/nearest/rint_test.cpp [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+
+#include <gtest/gtest.h>
+
+// Disabling test case ccm::nearbyintl if run on clang linux.
+#ifdef __clang__
+	#ifdef __linux__
+		#define CLANG_LINUX
+	#endif
+#endif
+
```
[the remark this answers is not in the document]
Comment (khurd21) [reply test/nearest/rint_test.cpp:24]: Looking at your other comments, I think I did make an oversight on the rounding modes. I was really only considering the compile time evaluation when I wrote them, so thanks for bringing this up!

I think my plan moving forward for this PR is to sometime this week / weekend, I will write tests that run for each rounding mode and then adapt what I currently have to work for the tests. I might be able to simply use your helper functions for the majority of these besides the rounding functions for long and long long types.

File: include/ccmath/math/nearest/rint.hpp (hunk 4 above)
Comment (khurd21) [reply include/ccmath/math/nearest/rint.hpp:136]: I will have to investigate this after writing the tests I think, cause I initially assumed this would be an OK cast to perform since the long and float are the same size? I will have to brush up on my casting :)

Thanks for pointing this out!

**Comment by khurd21** [comment]
@Rinzii 

I noticed two things from your comments that seem the most important. I wanted to write them down here because I don't think I addressed these in an earlier PR for implementing `ccm::nearbyint`.

- `constexpr` keyword being used to get the rounding mode: I used `constexpr` for `ccm::nearbyint` implementation.
- Testing rounding mode type at runtime. I did not add these for nearbyint, however it does look like adding them just for `rint` would be testing the same behaviors.

Should I also re-address these two things there? I think for this PR, I could change `constexpr` to `const` in the `ccm::nearbyint` and leave it at that, however it makes sense to test the rounding modes in that function as well as it would have caught this potential bug in getting the rounding mode.

Thanks for the review!

**Comment by khurd21** [comment]
@Rinzii 

I tried to address a few things in my recent changes:

- Removed the `constexpr` when getting the rounding mode to prevent compile time evaluation.
- Added a check for when `FE_INVALID` is raised to assert that it is being performed at runtime. To do this, I used `ccm::support::is_constant_evaluated`. Please do let me know if this was not was you were suggesting!

For addressing the different rounding modes, I was struggling to find a clean way to test them all without having a ton of duplicate code. I ended up making a decision to have cmake generate source files for us, changing the rounding mode for each source file generated. The end result is four generated files to test: `DOWNWARD`, `TONEAREST`, `TOWARDZERO`, and `UPWARD`. They all use the same template `rint_test.cpp.in`. This also forced me to separate out the tests that test for compile time evaluation. These are now in: `rint_test_compile_time.cpp`.

Here is an example of what they would look like in an IDE. For applications like VSCode, the files are placed in an out folder for the specific compiler in use.
Screenshot 2024-09-15 at 10 57 12 PM

In doing so, I also found that the `static_cast` for the `llrint`, `lrint` etc functions seem to be a non-issue. The rounding for the floating point values are already performed before the `static_cast` in performed, so because they are representing a whole number at that point, a static cast should not have an adverse effect on the number.

I did notice though that msvc handles `max` and `min` values slightly differently for `llrint` and `lrint` functions. If it is the maximum or minimum, it returns 0. Clang and gcc were not implemented this way. If this sounds OK to you, I was considering adding a check to return 0 for MSVC given these conditions.

**Comment by Rinzii** [comment]
Using cmake to generate different test cases is perfectly fine.

Also adding a msvc specific return is perfectly find and actually encouraged as ccmath tends to aim to follow the expected output of each compiler!

Also don't worry about taking your time with the PR! I'd rather you took the time to get everything right than rushing stuff!

If you have any specific questions for me just lmk! I've been pretty busy these last few weeks, but I'll try to get back to any questions you have promptly. :)

**Comment by Rinzii** [comment]
Hey @khurd21 sorry for being somewhat away. Been going through quite a bit recently, but I have made a discord specifically for ccmath to allow quicker collaboration between contributors or users of the library. Feel free to join if you so wish!

Discord Link:  https://discord.gg/p3mVxAbdmc

**Comment by Rinzii** [comment]
Also what is the current status of this PR if I may ask?

**Comment by khurd21** [comment]
> Also what is the current status of this PR if I may ask?

Hey! Sorry I haven't touched this in a long time! The current status is I believe I need to update the code slightly to be compatible with how MSVC handles the edge cases compared to other compilers. I also wanted to double check that the test cases were accurate

**Comment by Rinzii** [comment]
Hey @khurd21 I wanted to do one last check in. Is there still plans to continue this PR or should I close it?

**Comment by Rinzii** [comment]
Closing due to being stale.
